pandas/tests/frame/test_constructors.py

```diff
@@ -896,6 +896,46 @@ def test_constructor_list_of_dicts(self):
         expected = DataFrame(index=[0])
         assert_frame_equal(result, expected)
 
+    def test_constructor_ordered_dict_preserve_order(self):
+        # see gh-13304
+        expected = DataFrame([[2, 1]], columns=['b', 'a'])
+
+        data = OrderedDict()
+        data['b'] = [2]
+        data['a'] = [1]
+
+        result = DataFrame(data)
+        assert_frame_equal(result, expected)
+
+        data = OrderedDict()
+        data['b'] = 2
+        data['a'] = 1
+
+        result = DataFrame([data])
+        assert_frame_equal(result, expected)
+
+    def test_constructor_ordered_dict_conflicting_orders(self):
+        # the first element is an OrderedDict, so it sets the
+        # ordering for the DataFrame, even if there are conflicting
+        # orders from subsequent OrderedDict or dict objects
+        row_one = OrderedDict()
+        row_one['b'] = 2
+        row_one['a'] = 1
+
+        row_two = OrderedDict()
+        row_two['a'] = 1
+        row_two['b'] = 2
+
+        row_three = {'b': 2, 'a': 1}
+
+        expected = DataFrame([[2, 1], [2, 1]], columns=['b', 'a'])
+        result = DataFrame([row_one, row_two])
+        assert_frame_equal(result, expected)
+
+        expected = DataFrame([[2, 1], [2, 1], [2, 1]], columns=['b', 'a'])
+        result = DataFrame([row_one, row_two, row_three])
+        assert_frame_equal(result, expected)
+
```

Comment:
The test (at the end of test) I wrote was this:
~~~python
>>> expected = DataFrame([[1, 2], [1, 2]], columns=['a', 'b'])
>>> result = DataFrame([row_three, row_one])
>>> tm.assert_frame_equal(result, expected)
~~~
The keys ordering of `row_three` appears arbitrary and for some reason switches between the Python versions (2.x and 3.x).  So I'm leaving it out.